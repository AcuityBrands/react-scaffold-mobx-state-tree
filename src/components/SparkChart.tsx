import * as React from "react";
import * as Chart from "chart.js";
import { autorun } from "mobx";
import { observer } from "mobx-react";

interface ILineChartProps {
  id: string,
  label: string,
  height: number,
  renderSpeed:number,
}

// WARNING: This is a contrived example with limited utility

@observer
export default class SparkChart extends React.Component<ILineChartProps, null> {
  
  chartData: Array<number> = [
    12, 19, 3, 5, 2, 3, 8, 5, 4, 
    9, 1, 12, 3, 5, 0
  ];
  labels: Array<string> = [
    "1", "2", "3", "4", "5", "6", "7", "8", "9", 
    "10", "11", "12", "13", "14", "15"
  ];
  interval:any = -1;
  chart:any;
  disposer:any;
  
  componentDidMount() {
    this.renderChart();

    //Randomize the data, update if the rendering speed changes
    this.disposer = autorun(() => this.randomizeData(this.props.renderSpeed));
  }

  componentWillUnmount() {
    this.disposer();
  }

  renderChart() {
    //Chart globals
    Chart.defaults.global.legend.position = "bottom"
    Chart.defaults.global.elements.rectangle.borderWidth = 1;

    //Initialize Chart
    this.chart = new Chart(this.props.id, {
      type: 'line',
      data: {
        labels: this.labels,
        datasets: [{
          label: this.props.label,
          data: this.chartData,
          backgroundColor: this.getRandomColor(),
          pointRadius: 0,
          borderColor: 'rgba(0,0,0,0)'
        }]
      },
      options: {
        legend: { display: false },
        layout: {
          padding: {
            left: -10
          }
        },
        scales: {
          xAxes: [{
            display: false
          }],
          yAxes: [{
            ticks: { display: false },
            gridLines: {
                borderDash: [2,2],
                drawBorder: false
            }
          }]
        }
      }
    });
  }

  render() {
    return (
      <canvas id={this.props.id} height={this.props.height}></canvas>
    );
  }

  getRandomColor() {
    const o = Math.round, r = Math.random, s = 255;
    return 'rgba(' + o(r()*s) + ',' + o(r()*s) + ',' + o(r()*s) + ',0.5)';
  }

  randomizeData(value:number) {
    clearInterval(this.interval);

    if(value > 300) {
      this.chart.options.animation.duration = value;
    } else {
      this.chart.options.animation.duration = 0;
    }

    // Randomly update Chart
    this.interval = setInterval(() => {
      this.chartData.push(Math.floor(Math.random() * 10));
      this.chartData.shift();
      this.labels.push(Math.floor(Math.random() * 10).toString());
      this.labels.shift();
      this.chart.data.datasets[0].data = this.chartData;
      this.chart.update();
    }, value);
  }
}