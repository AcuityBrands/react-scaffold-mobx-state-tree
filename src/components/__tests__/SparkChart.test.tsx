import * as React from "react";
import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { SparkChart } from '../SparkChart';

Enzyme.configure({ adapter: new Adapter() });

const setup = propOverrides => {
  let spy
  const props = Object.assign({
    id: 'hello',
    label: 'test',
    height: 50,
    renderSpeed: 500
  }, propOverrides)
  const wrapper = Enzyme.shallow(<SparkChart {...props} />)
  return {
    props,
    wrapper
  }
}

describe('SparkChart Component', () => {
  beforeEach(() => {
    // Mock constructing the chart since we don't have full access to 
    // a canvas element with Jest.
    SparkChart.prototype.renderChart = jest.fn();
    SparkChart.prototype.randomizeData = jest.fn();
  });

  // Minimal component test confirms component rendered
  it("can render", () => {
    const { wrapper, props } = setup({});
    expect(wrapper.exists()).toBe(true);
  })

  it("matches the snapshot", () => {
    const { wrapper } = setup({});
    expect(wrapper).toMatchSnapshot();
  })

  it("will create a canvas element for the chart", () => {
    const { wrapper, props } = setup({});
    expect(wrapper.find('canvas').length).toBe(1);
  })

  it("will create and render the chart config on startup", () => {
    const spy = jest.spyOn(SparkChart.prototype, "renderChart")
    const { wrapper, props } = setup({});
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockClear();
  })
})