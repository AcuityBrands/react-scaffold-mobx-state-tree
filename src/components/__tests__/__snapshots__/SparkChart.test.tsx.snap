// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SparkChart Component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SparkChart
    height={50}
    id="hello"
    label="test"
    renderSpeed={500}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "height": 50,
      "id": "hello",
    },
    "ref": null,
    "rendered": null,
    "type": "canvas",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "height": 50,
        "id": "hello",
      },
      "ref": null,
      "rendered": null,
      "type": "canvas",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
